# Building a Racial Disparity Index for the Virginia Court Dataset
## 6/27/2022

In [1]:
from sqlalchemy import create_engine
import psycopg2 as db
import pandas as pd
import numpy as np
import os

In [2]:
postPass=os.environ["POSTGRES_PASS"]

In [3]:
engine = create_engine("postgresql+psycopg2://{user}:{pw}@localhost/{db}"
                       .format(user="jupyter", pw=postPass, db="expunge"))

## This is the 'census_long' table:

In [4]:
myquery = """
SELECT * from census_long
"""
pd.read_sql_query(myquery, con=engine)

FIPS         Jurisdiction  total_pop      race  population  proportion
0       1      Accomack County    32316.0     White     21899.0    0.677652
1       3     Albemarle County   109330.0     White     89388.0    0.817598
2       5     Alleghany County    14860.0     White     13783.0    0.927524
3       7        Amelia County    13145.0     White     10050.0    0.764549
4       9       Amherst County    31605.0     White     24299.0    0.768834
..    ...                  ...        ...       ...         ...         ...
793   800         Suffolk city    92108.0  Hispanic      4300.0    0.046684
794   810  Virginia Beach city   449974.0  Hispanic     38235.0    0.084972
795   820      Waynesboro city    22630.0  Hispanic      1966.0    0.086876
796   830    Williamsburg city    14954.0  Hispanic      1069.0    0.071486
797   840      Winchester city    28078.0  Hispanic      5127.0    0.182598

[798 rows x 6 columns]

## Here is the count of the number of cases within each FIPS/race/code section combination:

In [5]:
myquery = """
SELECT fips, race, code_section, COUNT(*) AS total_count_race 
FROM charges
GROUP BY fips, race, code_section
"""
pd.read_sql_query(myquery, con=engine)

fips                               race   code_section  \
0       001  American Indian or Alaskan Native       18.2-102   
1       001  American Indian or Alaskan Native       18.2-388   
2       001  American Indian or Alaskan Native        18.2-57   
3       001  American Indian or Alaskan Native       46.2-300   
4       001  American Indian or Alaskan Native       46.2-862   
...     ...                                ...            ...   
187104  840                              White         NO DMV   
187105  840                              White      Z.18.2-47   
187106  840                              White  Z.18.2-67.5:1   
187107  840                              White      Z.18.2-89   
187108  840                              White      Z.18.2-91   

        total_count_race  
0                      1  
1                      4  
2                      1  
3                      1  
4                      2  
...                  ...  
187104                10  
187105                 4  
187106                 1  
187107                 7  
187108                90  

[187109 rows x 4 columns]

## Here's the count by just FIPS/code section:

In [6]:
myquery = """
SELECT fips, code_section, COUNT(*) AS total_count_overall 
FROM charges
GROUP BY fips, code_section
"""
pd.read_sql_query(myquery, con=engine)

fips   code_section  total_count_overall
0       001         00-3.1                    1
1       001         002-06                    1
2       001         002-14                    1
3       001          003.1                    1
4       001           01-1                    1
...     ...            ...                  ...
109451  840  Z.18.2-67.5:1                    1
109452  840      Z.18.2-89                    7
109453  840      Z.18.2-90                    1
109454  840      Z.18.2-91                  116
109455  840      Z.18.2-92                    2

[109456 rows x 3 columns]

## Merging the previous two tables to get the fips/race/code section and fips/code section counts side by side:

In [7]:
myquery = """
SELECT c.fips, c.race, c.code_section, c.total_count_race, d.total_count_overall
FROM (SELECT fips, race, code_section, COUNT(*) AS total_count_race 
    FROM charges
    GROUP BY fips, race, code_section) c
INNER JOIN (SELECT fips, code_section, COUNT(*) AS total_count_overall 
    FROM charges
    GROUP BY fips, code_section) d
    ON c.fips = d.fips AND c.code_section=d.code_section
"""
pd.read_sql_query(myquery, con=engine)

fips                               race   code_section  \
0       001  American Indian or Alaskan Native       18.2-102   
1       001  American Indian or Alaskan Native       18.2-388   
2       001  American Indian or Alaskan Native       46.2-300   
3       001  American Indian or Alaskan Native       46.2-862   
4       001  American Indian or Alaskan Native     A.46.2-853   
...     ...                                ...            ...   
187104  840                              White     B.46.2-873   
187105  840                              White     B.46.2-878   
187106  840                              White     C.18.2-266   
187107  840                              White     C.46.2-862   
187108  840                              White  D.18.2-67.5:1   

        total_count_race  total_count_overall  
0                      1                   54  
1                      4                 1067  
2                      1                 5438  
3                      2                  116  
4                      1                  409  
...                  ...                  ...  
187104                 1                    1  
187105                 1                    1  
187106               183                  233  
187107               201                  287  
187108                 7                    7  

[187109 rows x 5 columns]

We can do this in a simpler way using the following code:

In [12]:
myquery = """
WITH byrace AS (
    SELECT fips, race, code_section, COUNT(*) AS total_count_race 
    FROM charges
    GROUP BY fips, race, code_section),
overall AS (
    SELECT fips, code_section, COUNT(*) AS total_count_overall 
    FROM charges
    GROUP BY fips, code_section)
SELECT c.fips, c.race, c.code_section, c.total_count_race, d.total_count_overall
FROM byrace c
INNER JOIN overall d
    ON c.fips = d.fips AND c.code_section=d.code_section
"""
pd.read_sql_query(myquery, con=engine)

fips      race code_section  total_count_race  total_count_overall
0       001     Black       00-3.1                 1                    1
1       001     White       002-06                 1                    1
2       001     White       002-14                 1                    1
3       001     Black        003.1                 1                    1
4       001     White         01-1                 1                    1
...     ...       ...          ...               ...                  ...
187104  840     Black    Z.18.2-90                 1                    1
187105  840     White    Z.18.2-91                90                  116
187106  840  Hispanic    Z.18.2-91                 3                  116
187107  840     Black    Z.18.2-91                23                  116
187108  840     Black    Z.18.2-92                 2                    2

[187109 rows x 5 columns]

## Joining both tables with census long:

In [15]:
myquery = """
WITH byrace AS (
    SELECT fips, race, code_section, COUNT(*) AS total_count_race 
    FROM charges
    GROUP BY fips, race, code_section),
overall AS (
    SELECT fips, code_section, COUNT(*) AS total_count_overall 
    FROM charges
    GROUP BY fips, code_section)
SELECT c.fips, c.race, c.code_section, c.total_count_race, d.total_count_overall
FROM byrace c
INNER JOIN overall d
    ON c.fips = d.fips AND c.code_section=d.code_section
INNER JOIN census_long l
    ON cast(c.fips as int) = l."FIPS" AND l.race=c.race
"""
pd.read_sql_query(myquery, con=engine)

fips      race code_section  total_count_race  total_count_overall
0       001     Black       00-3.1                 1                    1
1       001     White       002-06                 1                    1
2       001     White       002-14                 1                    1
3       001     Black        003.1                 1                    1
4       001     White         01-1                 1                    1
...     ...       ...          ...               ...                  ...
185595  840     Black    Z.18.2-90                 1                    1
185596  840     Black    Z.18.2-91                23                  116
185597  840     White    Z.18.2-91                90                  116
185598  840  Hispanic    Z.18.2-91                 3                  116
185599  840     Black    Z.18.2-92                 2                    2

[185600 rows x 5 columns]

## Generating the Disparity Index

In [46]:
myquery = """
WITH byrace AS (
    SELECT fips, race, code_section, COUNT(*) AS total_count_race 
    FROM charges
    GROUP BY fips, race, code_section),
overall AS (
    SELECT fips, code_section, COUNT(*) AS total_count_overall 
    FROM charges
    GROUP BY fips, code_section)
SELECT c.fips, l."Jurisdiction" as jurisdiction, c.race, c.code_section, c.total_count_race, d.total_count_overall, l.proportion,
    (cast(c.total_count_race as decimal)/cast(d.total_count_overall as decimal))/cast(l.proportion as decimal) as disparity_index
FROM byrace c
INNER JOIN overall d
    ON c.fips = d.fips AND c.code_section=d.code_section
INNER JOIN census_long l
    ON cast(c.fips as int) = l."FIPS" AND l.race=c.race
"""
pd.read_sql_query(myquery, con=engine)

In [47]:
#disp = pd.read_sql_query(myquery, con=engine)
#disp.to_sql("disparity", con=engine, index=False, if_exists='replace')

# Analysis:

## Sorting the Disparity Index and Filtering to Only Charges with > 1000 Occurrences:

In [49]:
myquery = """
SELECT * FROM disparity
WHERE total_count_overall > 1000
ORDER BY disparity_index DESC;
"""
pd.read_sql_query(myquery, con=engine)

fips           jurisdiction      race code_section  total_count_race  \
0     041    Chesterfield County   Unknown        13-60              1990   
1     081     Greensville County   Unknown     46.2-300               769   
2     035         Carroll County   Unknown   A.46.2-862               841   
3     113         Madison County   Unknown   A.46.2-862               164   
4     081     Greensville County   Unknown   A.46.2-862             12814   
...   ...                    ...       ...          ...               ...   
8564  013       Arlington County  Hispanic     18.2-388                 1   
8565  059         Fairfax County  Hispanic  18.2-108.01                 2   
8566  153  Prince William County  Hispanic     18.2-192                 1   
8567  059         Fairfax County  Hispanic     18.2-178                 2   
8568  153  Prince William County  Hispanic   18.2-258.1                 1   

      total_count_overall  proportion  disparity_index  
0                    2638    0.007208       104.655718  
1                    4345    0.002911        60.797043  
2                    4713    0.003793        47.044103  
3                    1256    0.003016        43.288296  
4                  102739    0.002911        42.844520  
...                   ...         ...              ...  
8564                 1437    0.156015         0.004460  
8565                 2969    0.165150         0.004079  
8566                 1114    0.244874         0.003666  
8567                 3512    0.165150         0.003448  
8568                 1231    0.244874         0.003317  

[8569 rows x 8 columns]

## Considering just African Americans:

In [51]:
myquery = """
SELECT * FROM disparity
WHERE total_count_overall > 1000 AND race='Black'
ORDER BY disparity_index DESC;
"""
pd.read_sql_query(myquery, con=engine)

fips       jurisdiction   race code_section  total_count_race  \
0     035     Carroll County  Black     46.2-300              1007   
1     035     Carroll County  Black   C.46.2-862              3058   
2     035     Carroll County  Black   18.2-250.1               256   
3     035     Carroll County  Black   A.46.2-862               677   
4     191  Washington County  Black   A.46.2-862              1150   
...   ...                ...    ...          ...               ...   
1493  027    Buchanan County  Black   B.46.2-301                 5   
1494  027    Buchanan County  Black     18.2-172                 4   
1495  027    Buchanan County  Black      18.2-95                 3   
1496  105         Lee County  Black     18.2-172                 4   
1497  105         Lee County  Black     18.2-456                 2   

      total_count_overall  proportion  disparity_index  
0                    5054    0.008828        22.569585  
1                   15878    0.008828        21.815794  
2                    1471    0.008828        19.713177  
3                    4713    0.008828        16.271237  
4                    6696    0.015184        11.310711  
...                   ...         ...              ...  
1493                 2109    0.033803         0.070135  
1494                 1733    0.033803         0.068282  
1495                 1347    0.033803         0.065887  
1496                 1688    0.037228         0.063652  
1497                 1020    0.037228         0.052669  

[1498 rows x 8 columns]

## Excluding traffic violations

In [52]:
myquery = """
SELECT * FROM disparity
WHERE total_count_overall > 1000 AND race='Black' AND code_section NOT LIKE '%%46.%%'
ORDER BY disparity_index DESC;
"""
pd.read_sql_query(myquery, con=engine)

fips       jurisdiction   race code_section  total_count_race  \
0     035     Carroll County  Black   18.2-250.1               256   
1     165  Rockingham County  Black     18.2-248              1040   
2     035     Carroll County  Black     18.2-172               106   
3     165  Rockingham County  Black     19.2-306               377   
4     013   Arlington County  Black     18.2-195               842   
...   ...                ...    ...          ...               ...   
1002  105         Lee County  Black     18.2-388                 3   
1003  027    Buchanan County  Black     18.2-172                 4   
1004  027    Buchanan County  Black      18.2-95                 3   
1005  105         Lee County  Black     18.2-172                 4   
1006  105         Lee County  Black     18.2-456                 2   

      total_count_overall  proportion  disparity_index  
0                    1471    0.008828        19.713177  
1                    3945    0.025406        10.376335  
2                    1396    0.008828         8.601016  
3                    1778    0.025406         8.345786  
4                    1063    0.096782         8.184366  
...                   ...         ...              ...  
1002                 1102    0.037228         0.073125  
1003                 1733    0.033803         0.068282  
1004                 1347    0.033803         0.065887  
1005                 1688    0.037228         0.063652  
1006                 1020    0.037228         0.052669  

[1007 rows x 8 columns]

Additional analyses:
* Generate bar chart of top 10 (or 15, 20) code sections by racial disparity for African American, Hispanic, White, etc ...
* Think about how to create an overall disparity score across all code sections within fips/race, then plot these scores on a map
* Create table of titles for each code section and join into existing data

Additional things to clean:
* Check on what race categories are matched or not matched
* Add others here as you go

In [5]:
myquery = """
SELECT * FROM disparity
WHERE total_count_overall > 1000 AND race='Hispanic' AND code_section NOT LIKE '%%46.%%'
ORDER BY disparity_index DESC;
"""
pd.read_sql_query(myquery, con=engine)

fips           jurisdiction      race code_section  total_count_race  \
0    059         Fairfax County  Hispanic     29.1-335               378   
1    003       Albemarle County  Hispanic   A.18.2-266               254   
2    171      Shenandoah County  Hispanic     18.2-248               259   
3    019         Bedford County  Hispanic     29.1-335                49   
4    067        Franklin County  Hispanic   A.18.2-266                71   
..   ...                    ...       ...          ...               ...   
907  013       Arlington County  Hispanic     18.2-388                 1   
908  059         Fairfax County  Hispanic  18.2-108.01                 2   
909  153  Prince William County  Hispanic     18.2-192                 1   
910  059         Fairfax County  Hispanic     18.2-178                 2   
911  153  Prince William County  Hispanic   18.2-258.1                 1   

     total_count_overall  proportion  disparity_index  
0                   1027    0.165150         2.228654  
1                   2489    0.057743         1.767309  
2                   2235    0.073528         1.576047  
3                   1572    0.023672         1.316778  
4                   1889    0.028782         1.305887  
..                   ...         ...              ...  
907                 1437    0.156015         0.004460  
908                 2969    0.165150         0.004079  
909                 1114    0.244874         0.003666  
910                 3512    0.165150         0.003448  
911                 1231    0.244874         0.003317  

[912 rows x 8 columns]

In [10]:
myquery = """
SELECT * FROM disparity
WHERE total_count_overall > 1000 AND jurisdiction = 'Fairfax County'
ORDER BY disparity_index DESC;
"""
pd.read_sql_query(myquery, con=engine)

fips    jurisdiction      race code_section  total_count_race  \
0    059  Fairfax County   Unknown   A.46.2-862              6775   
1    059  Fairfax County   Unknown   A.46.2-853               355   
2    059  Fairfax County   Unknown   C.46.2-862              3978   
3    059  Fairfax County   Unknown   B.46.2-853               109   
4    059  Fairfax County   Unknown       36-106               128   
..   ...             ...       ...          ...               ...   
405  059  Fairfax County  Hispanic        5-1-1                25   
406  059  Fairfax County  Hispanic     18.2-137                 4   
407  059  Fairfax County  Hispanic     18.2-192                 3   
408  059  Fairfax County  Hispanic  18.2-108.01                 2   
409  059  Fairfax County  Hispanic     18.2-178                 2   

     total_count_overall  proportion  disparity_index  
0                  56937    0.006187        19.231855  
1                   3371    0.006187        17.020647  
2                  47257    0.006187        13.605205  
3                   1303    0.006187        13.520369  
4                   2089    0.006187         9.903256  
..                   ...         ...              ...  
405                16917    0.165150         0.008948  
406                 3695    0.165150         0.006555  
407                 3534    0.165150         0.005140  
408                 2969    0.165150         0.004079  
409                 3512    0.165150         0.003448  

[410 rows x 8 columns]

example: notes: total pop of fairfax county = 1,337,047
hispanic pop = 189,515
14.2 % of pop is hisp
378 hisp charged with 29.1-335 (fishing w/o license) out of 1027 total charged in Fairfax 
hisp make up 36.8% of charged 
less than 1% of hisp pop are charged 

how to only show specific code sections to find specific code sections using disparity table? 